geopandas/tools/tests/test_clip.py

```diff
@@ -0,0 +1,322 @@
+
+
+def test_returns_gdf(point_gdf, single_rectangle_gdf):
+    """Test that function returns a GeoDataFrame (or GDF-like) object."""
+    out = clip(point_gdf, single_rectangle_gdf)
+    assert isinstance(out, GeoDataFrame)
+
+
+def test_returns_series(point_gdf, single_rectangle_gdf):
+    """Test that function returns a GeoSeries if GeoSeries is passed."""
+    out = clip(point_gdf.geometry, single_rectangle_gdf)
+    assert isinstance(out, GeoSeries)
+
+
+def test_non_overlapping_geoms():
+    """Test that a bounding box returns error if the extents don't overlap"""
+    unit_box = Polygon([(0, 0), (0, 1), (1, 1), (1, 0), (0, 0)])
+    unit_gdf = GeoDataFrame([1], geometry=[unit_box], crs={"init": "epsg:4326"})
+    non_overlapping_gdf = unit_gdf.copy()
+    non_overlapping_gdf = non_overlapping_gdf.geometry.apply(
+        lambda x: shapely.affinity.translate(x, xoff=20)
+    )
+    with pytest.raises(ValueError):
+        clip(unit_gdf, non_overlapping_gdf)
+
+
+def test_clip_points(point_gdf, single_rectangle_gdf):
+    """Test clipping a points GDF with a generic polygon geometry."""
+    clip_pts = clip(point_gdf, single_rectangle_gdf)
+    assert len(clip_pts.geometry) == 3 and clip_pts.geom_type[1] == "Point"
+
+
+def test_clip_poly(buffered_locations, single_rectangle_gdf):
+    """Test clipping a polygon GDF with a generic polygon geometry."""
+    clipped_poly = clip(buffered_locations, single_rectangle_gdf)
+    assert len(clipped_poly.geometry) == 3
+    assert all(clipped_poly.geom_type == "Polygon")
+
+
+def test_clip_poly_series(buffered_locations, single_rectangle_gdf):
+    """Test clipping a polygon GDF with a generic polygon geometry."""
+    clipped_poly = clip(buffered_locations.geometry, single_rectangle_gdf)
+    assert len(clipped_poly) == 3
+    assert all(clipped_poly.geom_type == "Polygon")
+
+
+def test_clip_multipoly_keep_slivers(multi_poly_gdf, single_rectangle_gdf):
+    """Test a multi poly object where the return includes a sliver.
+    Also the bounds of the object should == the bounds of the clip object
+    if they fully overlap (as they do in these fixtures)."""
+    with pytest.warns(UserWarning):
+        clipped = clip(multi_poly_gdf, single_rectangle_gdf)
+        warnings.warn("A geometry collection has been returned. ", UserWarning)
+        assert hasattr(clipped, "geometry")
+        assert np.array_equal(clipped.total_bounds, single_rectangle_gdf.total_bounds)
+        # Assert returned data is a geometry collection given sliver geoms
+        assert "GeometryCollection" in clipped.geom_type[0]
+
+
+def test_clip_multipoly_keep_geom_type(multi_poly_gdf, single_rectangle_gdf):
+    """Test a multi poly object where the return includes a sliver.
+    Also the bounds of the object should == the bounds of the clip object
+    if they fully overlap (as they do in these fixtures)."""
+    clipped = clip(multi_poly_gdf, single_rectangle_gdf, keep_geom_type=True)
+    assert hasattr(clipped, "geometry")
```

Comment:
General question: what is the `assert hasattr(clipped, "geometry")` supposed to test? Is it basically to test that the output is a GeoDataFrame? (but there is already a specific test for this I think)